### PR TITLE
Add Metastore TLS support for extended Hive connector implementations

### PR DESCRIPTION
**Pull request by dain**

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/HiveMetastoreClientFactory.java [hunk 1]
```diff
@@ -28,12 +29,18 @@
...
     private final HiveMetastoreAuthentication metastoreAuthentication;
 
-    public HiveMetastoreClientFactory(Optional<HostAndPort> socksProxy, Duration timeout, HiveMetastoreAuthentication metastoreAuthentication)
+    public HiveMetastoreClientFactory(
+            Optional<SSLContext> sslContext,
+            Optional<HostAndPort> socksProxy,
+            Duration timeout,
+            HiveMetastoreAuthentication metastoreAuthentication)
     {
+        this.sslContext = sslContext;
```
Comment (findepi) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/HiveMetastoreClientFactory.java]: `requireNN`

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/Transport.java [hunk 2]
```diff
@@ -47,44 +54,37 @@ public static TTransport create(HostAndPort address, Optional<HostAndPort> socks
...
-                socks.connect(InetSocketAddress.createUnresolved(address.getHost(), address.getPort()), timeoutMillis);
-                socks.setSoTimeout(timeoutMillis);
-                return new TSocket(socks);
+                socket.close();
             }
-            catch (Throwable t) {
-                closeQuietly(socks);
-                throw t;
+            catch (IOException e) {
+                t.addSuppressed(e);
```
Comment (findepi) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/Transport.java]: Looks like promoting `io.airlift.testing.Closeables#closeAllSuppress` to non-testing would be beneficial here (and not only here).

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/Transport.java [hunk 3]
```diff
@@ -47,44 +54,37 @@ public static TTransport create(HostAndPort address, Optional<HostAndPort> socks
...
         try {
+            // setSocketProperties(socket);
+            socket.connect(new InetSocketAddress(address.getHost(), address.getPort()), timeoutMillis);
+
+            if (sslContext.isPresent()) {
+                // SSL connect is to the socks address when present
+                HostAndPort sslConnectAddress = socksProxy.orElse(address);
+
+                socket = sslContext.get().getSocketFactory().createSocket(socket, sslConnectAddress.getHost(), sslConnectAddress.getPort(), true);
+                // setSocketProperties(socket);
```
Comment (findepi) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/Transport.java]: left?

(same 10 lines above)

**Comment by findepi** [comment]
What further steps are needed to use TLS when connecting to Metastore? Within this PR, this is a dead code, isn't it?

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/Transport.java [hunk 4]
```diff
@@ -47,44 +54,37 @@ public static TTransport create(HostAndPort address, Optional<HostAndPort> socks
...
-            return new TSocket(address.getHost(), address.getPort(), timeoutMillis);
-        }
+        Proxy proxy = socksProxy
+                .map(socksAddress -> new Proxy(SOCKS, InetSocketAddress.createUnresolved(socksAddress.getHost(), socksAddress.getPort())))
+                .orElse(Proxy.NO_PROXY);
 
-        Socket socks = createSocksSocket(socksProxy.get());
+        Socket socket = new Socket(proxy);
         try {
+            // setSocketProperties(socket);
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/Transport.java]: Remove

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/Transport.java [hunk 5]
```diff
@@ -47,44 +54,37 @@ public static TTransport create(HostAndPort address, Optional<HostAndPort> socks
...
+            socket.connect(new InetSocketAddress(address.getHost(), address.getPort()), timeoutMillis);
+
+            if (sslContext.isPresent()) {
+                // SSL connect is to the socks address when present
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/Transport.java]:
```java
// SSL will connect to the SOCKS address when present

**Comment by dain** [comment]
@findepi The metastore at FB uses TLS client certificates to authenticate (instead of kerberos).  I could add configuration to enable this for opensource, but I don't think anyone will use it.

**Pull request closed without merging**